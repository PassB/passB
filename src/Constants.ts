export enum executionContext {
  background = 'background',
  popup = 'popup',
  options = 'options',
}
