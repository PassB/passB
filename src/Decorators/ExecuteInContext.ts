// tslint:disable:no-any

import 'reflect-metadata';

export type ExecutionContext = 'background' | 'popup' | 'options' | undefined;
const executionContexts = ['background', 'popup'];

function getActionName(constructor: Function, propertyKey: string): string {
  return `${constructor.name}.${propertyKey}`;
}

interface AsynchronousRequest {
  action: string;
  params: any[];
}

export const executeInCorrectContext = () =>
  <R, T extends (this: any, ...args: any[]) => Promise<R>>
    (
      target: object,
      propertyKey: string,
      descriptor: TypedPropertyDescriptor<T>,
    ): TypedPropertyDescriptor<T> => {

    const executionContext = Reflect.getMetadata('executionContext', target, propertyKey);
    if (!executionContexts.includes(executionContext)) {
      throw new Error(`using ${executeInCorrectContext.name} `
        + 'decorator, but not declaring @Reflect.metadata("executionContext", "...") correctly',
      );
    }

    const wrappedFunction = descriptor.value || (() => Promise.reject(void 0));

    return {
      value: (async function wrapper(this: any, ...params: any[]): Promise<R> {
        if (window.executionContext === executionContext) {
          // console.debug('executing synchonously %s.%s', target.constructor.name, propertyKey);
          return await wrappedFunction.apply(this, params);
        }

        // console.debug('executing asynchonously %s.%s', target.constructor.name, propertyKey);
        return await browser.runtime.sendMessage({
          action: getActionName(target.constructor, propertyKey),
          params: [...params],
        });
      }) as T,
    };
  };

export const AsynchronousCallable = () =>
  <T extends { new(...args: any[]): {} }>(constructor: T): T =>
    class extends constructor {
      public constructor(...args: any[]) {
        super(...args);

        for (let obj = this; !!obj; obj = Object.getPrototypeOf(obj)) { // tslint:disable-line:no-this-assignment
          for (const propertyKey of Object.getOwnPropertyNames(obj)) {
            const executionContext = Reflect.getMetadata('executionContext', this, propertyKey);
            if (executionContext && window.executionContext === executionContext) {

              browser.runtime.onMessage.addListener(
                (request: AsynchronousRequest, sender: object, sendResponse: Function) => {
                  if (request.action !== getActionName(constructor, propertyKey)) {
                    return false;
                  }

                  const wrappedFn: ((...args: any[]) => Promise<any>) = (this as any)[propertyKey];
                  wrappedFn.apply(this, request.params).then(sendResponse);
                  return true;
                },
              );

            }
          }
        }
      }
    };

export const AsynchronousCallableServiceFactory =
  <X, T extends { new(...args: any[]): X }>(type: T): ((...args: any[]) => X) =>
    ((...args: any[]): X => (new (AsynchronousCallable()(type))(...args)));
