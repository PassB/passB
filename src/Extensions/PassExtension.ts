import {PassCli} from "../PassCli";
import {Extension, ListEntry, RegisterEntryCallback} from "./Extension";

export class PassExtension extends Extension {
  public name = 'Base';
  public actions = ['show'];

  public async initializeList(registerEntryCallback: RegisterEntryCallback): Promise<void> {
    for (const label of await PassCli.list()) {
      registerEntryCallback({
        label,
        actions: this.actions,
        metadata: {},
      });
    }
    return void 0;
  }

  public executeAction(action: string, entry: ListEntry): void {
    switch (action) {
      case 'show':
        break;
      default:
        console.error('unknown action:', action);
    }
  }
}
