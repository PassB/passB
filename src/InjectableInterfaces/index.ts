export {Extension} from './Extension';
export {FileFormat} from './FileFormat';
export {Filler} from './Filler';
export {Matcher} from './Matcher';
export {PassB} from './PassB';
export {PassCli} from './PassCli';
export {State} from './State';
export {Strategy} from './Strategy';
export {StorageAdaper} from './StorageAdapter';
