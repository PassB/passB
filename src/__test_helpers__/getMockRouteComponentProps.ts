import {RouteComponentProps} from 'react-router';

export function getMockRouteComponentProps<P>(data: P): RouteComponentProps<P> {
  const location = {
    hash: '',
    key: '',
    pathname: '',
    search: '',
    state: {},
  };
  return {
    match: {
      isExact: true,
      params: data,
      path: '',
      url: '',
    },
    location,
    history: {
      length: 2,
      action: 'POP',
      location,
      push: jest.fn(),
      replace: jest.fn(),
      go: jest.fn(),
      goBack: jest.fn(),
      goForward: jest.fn(),
      block: jest.fn(),
      createHref: jest.fn(),
      listen: jest.fn(),
    },
    staticContext: {},
  };
}
