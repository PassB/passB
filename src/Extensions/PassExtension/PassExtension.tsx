import * as React from 'react';
import {RouteProps} from "react-router";
import {Service} from 'typedi';
import {executeInCorrectContext, AsynchronousCallableServiceFactory} from "Decorators/ExecuteInContext";
import {LazyInject} from "Decorators/LazyInject";
import {OptionsPanelType} from "Options/OptionsReceiver";
import {PassB} from "PassB";
import {PassCli} from "PassCli";
import {ExecutionOptions, Extension, ExtensionTag, RegisterEntryCallback} from "..";
import {Show} from "./Views";

interface Options {
}

@Service({
  tags: [ExtensionTag],
  factory: AsynchronousCallableServiceFactory(PassExtension),
})
export class PassExtension extends Extension<Options> {
  public static readonly routes: RouteProps[] = [
    {
      path: '/extension/Pass/Show',
      component: Show,
    },
  ];
  public readonly name: string = 'Pass';
  public readonly actions: string[] = ['show', 'fill'];
  public readonly defaultOptions: Options = {};
  public readonly OptionsPanel?: OptionsPanelType<Options> = void 0;

  @LazyInject(() => PassB)
  private passB: PassB;

  @LazyInject(() => PassCli)
  private passCli: PassCli;

  public async initializeList(registerEntryCallback: RegisterEntryCallback): Promise<void> {
    for (const label of await this.passCli.list()) {
      registerEntryCallback({
        label,
        actions: this.actions,
      });
    }
  }

  public getLabelForAction(action: string): string {
    switch (action) {
      case 'show':
        return 'extension_pass_action_show';
      case 'fill':
        return 'extension_pass_action_fill';
      default:
        return '';
    }
  }

  public executeAction(action: string, entry: string, {navigateTo}: ExecutionOptions): void {
    switch (action) {
      case 'show':
        navigateTo('/extension/Pass/Show', {entry});
        break;
      case 'fill':
        this.executeFillAction(entry);
        window.close();
        break;
      default:
        console.error('unknown action:', action);
    }
  }

  private async getCurrentTab(): Promise<browser.tabs.Tab> {
    return (await browser.tabs.query({active: true, currentWindow: true}))[0];
  }

  @executeInCorrectContext()
  @Reflect.metadata("executionContext", "background")
  private async executeFillAction(entry: string): Promise<{}> {
    const initialUrl = (await this.getCurrentTab()).url;
    const entryContents = await this.passCli.show(entry);
    const activeTab = await this.getCurrentTab();
    const finalUrl = activeTab.url;

    if (typeof finalUrl !== "undefined" && initialUrl !== finalUrl) {
      console.info(`url changed from request to receive of password. not filling`);
      return {};
    }

    const filler = await this.passB.getFiller();
    const fileFormat = await this.passB.getFileFormat();

    return Promise.all([
      filler.fillPassword(activeTab, fileFormat.getPassword(entryContents, entry)),
      filler.fillUsername(activeTab, fileFormat.getUsername(entryContents, entry)),
    ]);
  }
}
