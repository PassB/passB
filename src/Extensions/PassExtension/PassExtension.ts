import {injectable} from 'inversify';
import {RouteProps} from 'react-router';
import {executionContext} from 'Constants';
import {Interfaces, Symbols} from 'Container';
import {lazyInject} from 'Decorators/lazyInject';
import {executeInCorrectContext, AsynchronousCallable} from 'Decorators/ExecuteInContext';
import {ExecutionOptions} from 'InjectableInterfaces/Extension';
import {OptionsPanelType} from 'InjectableInterfaces/OptionsPanel';
import {EntryAction} from 'State/PassEntries/Interfaces';
import {createTypedMap} from 'State/Types/TypedMap';
import {Extension} from '..';
import {Show} from './Views';

interface Options {
}

@AsynchronousCallable()
@injectable()
export class PassExtension extends Extension<Options> {
  public readonly routes: RouteProps[] = [
    {
      path: '/extension/Pass/Show',
      component: Show,
    },
  ];
  public readonly actions: string[] = ['show', 'fill'];
  public readonly OptionsPanel?: OptionsPanelType<Options> = void 0;

  @lazyInject(Symbols.PassB)
  private passB: Interfaces.PassB;

  @lazyInject(Symbols.PassCli)
  private passCli: Interfaces.PassCli;

  public constructor() {
    super('Pass', createTypedMap({}));
  }

  public async initializeList(): Promise<void> {
    const entries: EntryAction[] = [];

    for (const fullPath of await this.passCli.list()) {
      if (fullPath === '' || fullPath.endsWith('/')) {
        continue;
      }

      entries.push(...this.actions.map((action: string) => ({fullPath, action})));
    }
    this.setEntries(entries);
  }

  public getLabelForAction(action: string): string {
    switch (action) {
      case 'show':
        return 'extension_pass_action_show';
      case 'fill':
        return 'extension_pass_action_fill';
      default:
        return '';
    }
  }

  public executeAction(action: string, entry: string, {navigateTo}: ExecutionOptions): void {
    switch (action) {
      case 'show':
        navigateTo('/extension/Pass/Show', {entry});
        break;
      case 'fill':
        this.executeFillAction(entry);
        window.close();
        break;
      default:
        console.error('unknown action:', action);
        break;
    }
  }

  private async getCurrentTab(): Promise<browser.tabs.Tab> {
    return (await browser.tabs.query({active: true, currentWindow: true}))[0];
  }

  @executeInCorrectContext(executionContext.background)
  private async executeFillAction(entry: string): Promise<{}> {
    const initialUrl = (await this.getCurrentTab()).url;
    const entryContents = await this.passCli.show(entry);
    const activeTab = await this.getCurrentTab();
    const finalUrl = activeTab.url;

    if (typeof finalUrl !== 'undefined' && initialUrl !== finalUrl) {
      console.info('url changed from request to receive of password. not filling');
      return {};
    }

    const filler = await this.passB.getFiller();
    const fileFormat = await this.passB.getFileFormat();

    return Promise.all([
      filler.fillPassword(activeTab, fileFormat.getPassword(entryContents, entry)),
      filler.fillUsername(activeTab, fileFormat.getUsername(entryContents, entry)),
    ]);
  }
}
