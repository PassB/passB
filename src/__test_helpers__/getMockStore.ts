import {combineReducers, createStore, Store} from 'redux';
import {StoreContents} from 'InjectableInterfaces/State';
import {initialState as initialOptionsState, reducer as OptionsReducer} from 'State/Options/Reducers';
import {initialState as initialPassEntriesState, reducer as PassEntryReducer} from 'State/PassEntries/Reducers';

const reducer = combineReducers<StoreContents>({
  options: OptionsReducer,
  passEntries: PassEntryReducer,
});

export const getMockStore = (): Store<StoreContents> => createStore<StoreContents>(
  reducer,
  {
    options: initialOptionsState,
    passEntries: initialPassEntriesState,
  },
);
