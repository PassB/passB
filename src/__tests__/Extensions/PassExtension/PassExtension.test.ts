import {getMockStore} from '__test_helpers__/getMockStore';
import {container, Symbols} from 'Container';
import {setExecutionContext} from 'Decorators/ExecuteInContext';
import {PassExtension} from 'Extensions/PassExtension/PassExtension';
import {Extension} from 'InjectableInterfaces';
import {executionContext} from '../../../Constants';

describe('PassExtension', () => {
  beforeEach(() => {
    setExecutionContext('test');
    container.snapshot();
  });

  afterEach(() => {
    container.restore();
  });

  it('is bound as Symbols.Extension', () => {
    expect(
      container.getAll(Symbols.Extension)
        .some((extension: Extension<{}>) => extension instanceof PassExtension),
    ).toBeTruthy();
  });

  it('is named "Pass"', () => {
    expect(container.resolve(PassExtension).name).toBe('Pass');
  });

  it('has translated labels for all actions', () => {
    const extension = container.resolve(PassExtension);
    for (const action of extension.actions) {
      const label = extension.getLabelForAction(action);
      expect(label).not.toBeUndefined();
      expect(browser.i18n.getMessage(label!)).not.toBeUndefined();
    }
  });

  describe('initializeList method', () => {
    beforeEach(() => {
      container.rebind(Symbols.State).toConstantValue({getStore: jest.fn(getMockStore)});
    });

    it('calls PassCli.list', async () => {
      const list = jest.fn(() => Promise.resolve([]));
      container.rebind(Symbols.PassCli).toConstantValue({list});
      await container.resolve(PassExtension).initializeList();
      expect(list).toHaveBeenCalled();
    });

    it('registers actions only on files, not on folders', async () => {
      const list = jest.fn(() => Promise.resolve([
        'folder/', 'sub/folder/', 'file', 'sub/file',
      ]));
      container.rebind(Symbols.PassCli).toConstantValue({list});
      const extension = container.resolve(PassExtension);

      const spy = jest.spyOn(extension as any, 'setEntries'); // tslint:disable-line:no-any

      await extension.initializeList();

      expect(spy).toHaveBeenCalled();
      expect(spy.mock.calls).toMatchSnapshot();
    });
  });

  describe('actions', () => {
    describe('show', () => {
      it('starts navigation to /extension/Pass/Show with entry name as parameter', () => {
        const navigateTo = jest.fn();
        container.resolve(PassExtension).executeAction('show', 'my/test/entry', {navigateTo});
        expect(navigateTo).toHaveBeenCalledTimes(1);
        expect(navigateTo.mock.calls[0]).toEqual(['/extension/Pass/Show', {entry: 'my/test/entry'}]);
      });
    });

    describe('fill', () => {
      let tabSpy: jest.Mock<Promise<[{url: string}]>>;
      let passCli: { show: jest.Mock<Promise<string[]>> };
      let fileFormat: { getUsername: jest.Mock<string>; getPassword: jest.Mock<string> };
      let filler: { fillUsername: jest.Mock<Promise<void>>; fillPassword: jest.Mock<Promise<void>> };

      beforeEach(() => {
        tabSpy = ((global as any).browser.tabs.query as jest.Mock<Promise<[{url: string}]>>); // tslint:disable-line:no-any
        tabSpy.mockReset();
        tabSpy.mockReturnValue(Promise.resolve([{url: 'https://example.com'}]));

        passCli = {show: jest.fn(() => Promise.resolve(['firstLine', 'secondLine']))};
        container.rebind(Symbols.PassCli).toConstantValue(passCli);

        fileFormat = {getUsername: jest.fn(() => 'user'), getPassword: jest.fn(() => 'pass')};
        filler = {fillPassword: jest.fn(() => Promise.resolve()), fillUsername: jest.fn(() => Promise.resolve())};
        container.rebind(Symbols.PassB).toConstantValue({
          getFileFormat: () => fileFormat,
          getFiller: () => filler,
        });
      });

      it('calls executeAction and then closes the window', () => {
        const instance = container.resolve(PassExtension);
        const executeFillActionSpy = jest.spyOn(instance as any, 'executeFillAction'); // tslint:disable-line:no-any
        const windowCloseSpy = jest.spyOn(window, 'close');

        instance.executeAction('fill', 'my/test/entry', {navigateTo: jest.fn()});
        expect(executeFillActionSpy).toHaveBeenCalledWith('my/test/entry');
        expect(windowCloseSpy).toHaveBeenCalled();
      });

      it('normal filling', async () => {
        setExecutionContext(executionContext.background);
        await (container.resolve(PassExtension) as any).executeFillAction('my/test/entry'); // tslint:disable-line:no-any

        expect(tabSpy).toHaveBeenCalledTimes(2);
        expect(passCli.show).toHaveBeenCalledWith('my/test/entry');
        expect(fileFormat.getPassword).toHaveBeenCalledWith(['firstLine', 'secondLine'], 'my/test/entry');
        expect(fileFormat.getUsername).toHaveBeenCalledWith(['firstLine', 'secondLine'], 'my/test/entry');
        expect(filler.fillUsername).toHaveBeenCalledWith({url: 'https://example.com'}, 'user');
        expect(filler.fillPassword).toHaveBeenCalledWith({url: 'https://example.com'}, 'pass');
      });

      it('does not fill when url changes while pass contents are received', async () => {
        setExecutionContext(executionContext.background);

        tabSpy
          .mockReturnValueOnce(Promise.resolve([{url: 'https://example.com'}]))
          .mockReturnValueOnce(Promise.resolve([{url: 'https://example.com/attack'}]))
        ;

        const spy = jest.spyOn(console, 'warn').mockImplementationOnce(() => 0);

        await (container.resolve(PassExtension) as any).executeFillAction('my/test/entry'); // tslint:disable-line:no-any

        expect(spy).toHaveBeenCalled();

        expect(tabSpy).toHaveBeenCalledTimes(2);
        expect(passCli.show).toHaveBeenCalledWith('my/test/entry');
        expect(fileFormat.getPassword).not.toHaveBeenCalled();
        expect(fileFormat.getUsername).not.toHaveBeenCalled();
        expect(filler.fillUsername).not.toHaveBeenCalled();
        expect(filler.fillPassword).not.toHaveBeenCalled();
      });
    });
  });
});
