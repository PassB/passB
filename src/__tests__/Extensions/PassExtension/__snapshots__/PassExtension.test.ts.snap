// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PassExtension initializeList method registers actions only on files, not on folders 1`] = `
Array [
  Array [
    Array [
      Object {
        "action": "show",
        "fullPath": "file",
      },
      Object {
        "action": "fill",
        "fullPath": "file",
      },
      Object {
        "action": "show",
        "fullPath": "sub/file",
      },
      Object {
        "action": "fill",
        "fullPath": "sub/file",
      },
    ],
  ],
]
`;
